README.md
# Cylon.js For Powerup

Cylon.js (http://cylonjs.com) is a JavaScript framework for robotics, physical computing, and the Internet of Things (IoT).

This repository contains the Cylon driver for the [Powerup 3.0](http://www.poweruptoys.com/products/powerup-v3) Bluetooth LE controlled paper airplane.

Want to use Ruby on robots? Check out our sister project Artoo (http://artoo.io)

Want to use the Go programming language to power your robots? Check out our sister project Gobot (http://gobot.io).

[![Build Status](https://secure.travis-ci.org/hybridgroup/cylon-powerup.png?branch=master)](http://travis-ci.org/hybridgroup/cylon-powerup) [![Code Climate](https://codeclimate.com/github/hybridgroup/cylon-powerup/badges/gpa.svg)](https://codeclimate.com/github/hybridgroup/cylon-powerup) [![Test Coverage](https://codeclimate.com/github/hybridgroup/cylon-powerup/badges/coverage.svg)](https://codeclimate.com/github/hybridgroup/cylon-powerup)

## How to Install

Install the module with:

    $ npm install cylon cylon-powerup

## How to Use

Here's an example of flying the powerup with a dualshock3 controller:

```javascript
var Cylon = require('cylon');

Cylon.robot({
  connections: {
    bluetooth: { adaptor: 'ble', uuid: '84dd20eb3d89' },
    joystick: { adaptor: 'joystick' }
  },

  devices: {
    controller: { driver: 'dualshock-3', connection: 'joystick' },
    powerup: { driver: 'powerup' }
  },

  work: function(my) {
    var thrust = 0;
    var rudder = 0;
    var canRudder = false;

    var cb = function(err) {
      if (!!err) {
        console.log(err);
      } else {
        if (canRudder) {
          canRudder = false;
          my.powerup.setRudder(rudder, cb);
        }
        else {
          my.powerup.setThrust(thrust, cb);
        }
      }
    }

    my.powerup.setThrust(thrust, cb);

    my.controller.on("left_y:move", function(data) {
      if (data < 0) {
        thrust = Math.abs(data/32768*254) | 0;
      } else {
        thrust = 0;
      }
    });

    my.controller.on("right_x:move", function(data) {
      var tmp = data/32768*127 | 0;
      if (tmp !== rudder) {
        rudder = tmp;
        canRudder = true;
      }
    });
  }
}).start();
```

## How to Connect

The Powerup is a Bluetooth Low-Energy device, and is connected as with any other BLE device. For more info, check out the [BLE platform page](http://cylonjs.com/documentation/platforms/ble).

## Documentation

We're busy adding documentation to our web site at http://cylonjs.com/ please check there as we continue to work on Cylon.js

Thank you!

## Contributing

For our contribution guidelines, please go to [https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md
](https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md
).

## Release History

For the release history, please go to [https://github.com/hybridgroup/cylon-powerup/blob/master/RELEASES.md
](https://github.com/hybridgroup/cylon-powerup/blob/master/RELEASES.md
).

## License

Copyright (c) 2014-2015 The Hybrid Group. Licensed under the Apache 2.0 license.


RELEASES.md
## Release History

Version 0.7.0 - Use Cylon 1.2.0 & cylon-ble 0.8.0

Version 0.6.0 - Use Cylon 1.1.0

Version 0.5.0 - Compatability with Cylon 1.0

Version 0.4.0 - Compatability with Cylon 0.22.0

Version 0.3.0 - Compatability with Cylon 0.21.0

Version 0.2.0 - Compatability with Cylon 0.20.0

Version 0.1.0 - Initial release


spec/lib/driver.spec.js
"use strict";

var Powerup = lib("driver");

describe("Cylon.Drivers.Powerup", function() {
  var driver = new Powerup({
    device: { connection: "connect" }
  });

  it("needs tests", function() {
    expect(driver).to.be.an.instanceOf(Powerup);
  });
});


test/specs/adaptor.spec.js
'use strict';

var Powerup = source("adaptor");

describe("Cylon.Adaptors.Powerup", function() {
  var adaptor = new Powerup();

  it("needs tests");
});


spec/lib/cylon-powerup.spec.js
"use strict";

var powerup = lib("../");

var Driver = lib("driver");

describe("Cylon.Powerup", function() {
  describe("#drivers", function() {
    it("is an array of provided drivers", function() {
      expect(powerup.drivers).to.be.eql(["powerup"]);
    });
  });

  describe("#driver", function() {
    it("returns an instance of the Driver", function() {
      var args = { adaptor: {} };
      expect(powerup.driver(args)).to.be.instanceOf(Driver);
    });
  });
});


lib/driver.js
/*
 * cylon-powerup driver
 * http://cylonjs.com
 *
 * Copyright (c) 2014 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

"use strict";

var Cylon = require("cylon");

var MOTOR = "75b64e51f1844ed1921a476090d80ba7",
    RUDDER = "75b64e51f1854ed1921a476090d80ba7";

var Driver = module.exports = function Driver() {
  Driver.__super__.constructor.apply(this, arguments);

  this.serviceId = "75b64e51f1814ed1921a476090d80ba7";

  this.commands = {
    setThrust: this.setThrust,
    setRudder: this.setRudder
  };
};

Cylon.Utils.subclass(Driver, Cylon.Driver);

/**
 * Starts the driver
 *
 * @param {Function} callback to be triggered when started
 * @return {void}
 */
Driver.prototype.start = function(callback) {
  callback();
};

/**
 * Stops the driver
 *
 * @param {Function} callback to be triggered when halted
 * @return {void}
 */
Driver.prototype.halt = function(callback) {
  callback();
};

/**
 * Sets the thrust value of the Powerup
 *
 * @param {Number} value the new value for the Powerup's thrust
 * @param {Function} callback to be triggered when the value is written
 * @return {void}
 * @publish
 */
Driver.prototype.setThrust = function(value, callback) {
  this._writeServiceCharacteristic(value, MOTOR, callback);
};

/**
 * Sets the rudder value of the Powerup
 *
 * @param {Number} value the new value for the Powerup's rudder
 * @param {Function} callback to be triggered when the value is written
 * @return {void}
 * @publish
 */
Driver.prototype.setRudder = function(value, callback) {
  this._writeServiceCharacteristic(value, RUDDER, callback);
};

Driver.prototype._writeServiceCharacteristic = function(value, char, callback) {
  this.connection.writeServiceCharacteristic(
    this.serviceId,
    char,
    new Buffer([value]),
    function(err) {
      if (typeof callback === "function") { callback(err); }
    }
  );
};


examples/dualshock3/dualshock3.js
"use strict";

var Cylon = require("cylon");

Cylon.robot({
  connections: {
    bluetooth: { adaptor: "ble", uuid: "84dd20eb3d89" },
    dualshock3: { adaptor: "joystick" }
  },

  devices: {
    controller: { driver: "dualshock-3", connection: "dualshock3" },
    powerup: { driver: "powerup" }
  },

  work: function(my) {
    var thrust = 0;
    var rudder = 0;
    var canRudder = false;

    var cb = function(err) {
      if (err) {
        return console.log(err);
      }

      if (canRudder) {
        canRudder = false;
        my.powerup.setRudder(rudder, cb);
      } else {
        my.powerup.setThrust(thrust, cb);
      }
    };

    my.powerup.setThrust(thrust, cb);

    my.controller.on("left_y:move", function(data) {
      if (data < 0) {
        thrust = Math.abs(data / 32768 * 254) | 0;
      } else {
        thrust = 0;
      }
    });

    my.controller.on("right_x:move", function(data) {
      var tmp = data / 32768 * 127 | 0;
      if (tmp !== rudder) {
        rudder = tmp;
        canRudder = true;
      }
    });
  }
}).start();


examples/powerup_dualshock3.js
var Cylon = require('cylon');

Cylon.robot({
  connections: [
    { name: 'bluetooth', adaptor: 'ble', uuid: '84dd20eb3d89' },
    { name: 'dualshock3', adaptor: 'joystick' }
  ],
  devices: [
    { name: 'controller', driver: 'dualshock-3', connection: 'dualshock3' }, 
    { name: 'powerup', driver: 'powerup' }
  ],
  work: function(my) {
    var thrust = 0;
    var rudder = 0;
    var canRudder = false;

    var cb = function(err) {
      if (!!err) {
        console.log(err);
      } else {
        if (canRudder) {
          canRudder = false;
          my.powerup.setRudder(rudder, cb);
        }
        else {
          my.powerup.setThrust(thrust, cb);
        }
      }
    }

    my.powerup.setThrust(thrust, cb);

    my.controller.on("left_y:move", function(data) {
      if (data < 0) {
        thrust = Math.abs(data/32768*254) | 0;
      } else {
        thrust = 0;
      }
    });

    my.controller.on("right_x:move", function(data) {
      var tmp = data/32768*127 | 0;
      if (tmp !== rudder) {
        rudder = tmp;
        canRudder = true;
      }
    });
  }
}).start();
